src/os/elevate.cpp

```diff
@@ -0,0 +1,155 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "os/elevate.h"
+
+static std::string JoinArguments(std::vector<std::string> arguments) {
+  std::ostringstream result;
+
+  std::copy(arguments.begin(), arguments.end(),
+            std::ostream_iterator<std::string>(result, " "));
+
+  return result.str();
+}
+
+// Make sure to delete the result after you're done
+// with it by calling `delete[] result;`.
+// See http://stackoverflow.com/a/1201471
+static LPCTSTR ConvertStringToLPCTSTR(const std::string &string) {
+  char *result = new char[string.size() + 1];
+  std::copy(string.begin(), string.end(), result);
+  result[string.size()] = 0;
+  return result;
+}
+
+etcher::ELEVATE_RESULT etcher::Elevate(const std::string &command,
+                                       std::vector<std::string> arguments) {
+  // Initialize the SHELLEXECUTEINFO structure. We zero it out
+  // in order to be on the safe side, and set cbSize to the size
+  // of the structure as recommend by MSDN
+  // See: https://msdn.microsoft.com/en-us/library/windows/desktop/bb759784(v=vs.85).aspx
+  SHELLEXECUTEINFO shellExecuteInfo;
+  ZeroMemory(&shellExecuteInfo, sizeof(shellExecuteInfo));
+  shellExecuteInfo.cbSize = sizeof(SHELLEXECUTEINFO);
+
+  // Flags that indicate the content and validity of the other structure member.
+  shellExecuteInfo.fMask =
+
+      // Used to indicate that the hProcess member receives the process handle.
+      // This handle is typically used to allow an application to find out
+      // when a process created with ShellExecuteEx terminates.
+      SEE_MASK_NOCLOSEPROCESS |
+
+      // Wait for the execute operation to complete before returning.
+      SEE_MASK_NOASYNC |
+
+      // Do not display an error message box if an error occurs.
+      SEE_MASK_FLAG_NO_UI;
+
+  // The action to be performed.
+  shellExecuteInfo.lpVerb = TEXT("runas");
+
+  // Run the file in the background
+  shellExecuteInfo.nShow = SW_HIDE;
+
+  // Use the current directory as the working directory
+  shellExecuteInfo.lpDirectory = NULL;
+
+  // Set file and parameters
+  // We can't just assign the result of `.c_str()`, since
+  // that pointer is owned by the `std::string` instance,
+  // and will not be safe after the instance is destroyed.
+  LPCTSTR file = ConvertStringToLPCTSTR(command);
+  LPCTSTR argv = ConvertStringToLPCTSTR(JoinArguments(arguments));
+  shellExecuteInfo.lpFile = file;
+  shellExecuteInfo.lpParameters = argv;
+
+  BOOL executeResult = ShellExecuteEx(&shellExecuteInfo);
+
+  delete[] file;
+  delete[] argv;
+
+  // Finally, let's try to elevate the command
+  if (!executeResult) {
+    DWORD executeError = GetLastError();
+
+    switch (executeError) {
+    case ERROR_FILE_NOT_FOUND:
+      return etcher::ELEVATE_RESULT::ELEVATE_FILE_NOT_FOUND;
```

Comment:
I don't want to return strings from this function, but constants that we can easily check for later on. We could make this function return a `DWORD`, however not all errors coming from the Windows API are represented using `DWORD`, so its nice to have an extra layer (the `enum class`) for normalization purposes (both to encapsulate Windows specific details and present a uniform error interface).